ext/debug/ctf/README.rst

```diff
@@ -0,0 +1,216 @@
+.. _ctf:
+
+Common Trace Format for Zephyr
+##############################
+
+Common Trace Format, CTF, is an open format and language to describe trace
+formats.  This enables tool reuse, of which line-textual (babeltrace) and
+graphical (TraceCompass) variants already exist.
+
+CTF should look familiar to C programmers but adds stronger typing.
+See `CTF - A  Flexible, High-performance Binary Trace Format
+<http://diamon.org/ctf/>`_.
+
+Every system has application-specific events to trace out.  Historically,
+that has implied:
+
+1. Determining the application-specific payload,
+2. Choosing suitable serialization-format,
+3. Writing the on-target serialization code,
+4. Deciding on and writing the I/O transport mechanics,
+5. Writing the PC-side deserializer/parser,
+6. Writing custom ad-hoc tools for filtering and presentation.
+
+CTF allows us to formally describe #1 and #2, which enables common
+infrastructure for #5 and #6.  This leaves #3 serialization code and #4
+I/O mechanics up to a custom implementation.
+
+This CTF debug module aims at providing a common #1 and #2 for Zephyr
+("middle"), while providing a lean & generic interface for I/O ("bottom").
+Currently, only one CTF bottom-layer exists, POSIX ``fwrite``, but many others
+are possible:
+
+- Async UART
+- Async DMA
+- Sync GPIO
+- ... and many more.
+
+In fact, I/O varies greatly from system to system.  Therefore, it is
+instructive to create a taxonomy for I/O types when we must ensure the
+interface between CTF-middle and CTF-bottom is generic and efficient
+enough to model these. See the *I/O taxonomy* section below.
+
+
+A generic interface
+-------------------
+
+In CTF, an event is serialized to a packet containing one or more fields
+and as seen from *I/O taxonomy* section below, a bottom layer:
+
+- may perform action at transaction-start (e.g. mutex-lock),
+- may process each field in some way (e.g. sync-push emit, concat, enqueue to thread-bound FIFO),
+- may perform action at transaction-stop (e.g. mutex-release, emit of concat buffer).
+
+This is possible when the bottom-layer implements the following macros:
+
+- ``CTF_BOTTOM_LOCK``:   No-op or how to lock the I/O transaction.
+- ``CTF_BOTTOM_UNLOCK``: No-op or how to release the I/O transaction.
+- ``CTF_BOTTOM_FIELDS``: Var-args of fields. May process each field with ``MAP``.
+- ``CTF_BOTTOM_TIMESTAMPED_INTERNALLY``: Tells where timestamping is done.
+
+These macros along with inline functions of the middle-layer can yield
+very low-overhead tracing infrastructure.
+
+
+CTF middle-layer example
+------------------------
+
+The CTF_EVENT macro will serialize each argument to a field:
```

Comment:
If you end this line with a double colon (``::``) the following block of indented text will be treated as a code-block (which is probably what you intended).

Reply:
Fixed, thanks!